File: packages/core/src/artifact-files.ts
import fs from "node:fs";
import path from "node:path";

/**
 * The filesystem view of one change directory: which files are artifacts, their kinds, their names,
 * and their mtimes. Everything here reads directory entries and stats only. It does not read file
 * content. discoverArtifacts in artifact-discovery.ts builds the ChangeArtifact objects on top of this list.
 */

/** The root-level extensions of a data artifact. Exported so a file watcher can build the set of
 *  extensions worth reacting to (`.md` + these) from one source. Re-listing them would drift when a new
 *  data extension is added here. */
export const DATA_EXTENSIONS = [".yaml", ".yml", ".json"];

/** The root-level extensions of a diagram artifact. Exported for the same reason as DATA_EXTENSIONS: a
 *  file watcher builds the set of extensions worth reacting to from these two lists, not from a third
 *  copy that drifts when one is extended. A diagram is its own kind rather than another data extension —
 *  `data` means "show this file's text", `diagram` means "show what this file describes", and folding the
 *  two together makes the raw source the only rendering a diagram file can ever get. */
export const DIAGRAM_EXTENSIONS = [".mmd", ".mermaid"];

/** The kind of a root-level artifact file. specs is not here: it is a tree, not a root file. */
export type RootKind = "tasks" | "markdown" | "data" | "diagram";

const isTasksName = (n: string) => n === "tasks.md";
const isMarkdownName = (n: string) => n.endsWith(".md");
const isDataName = (n: string) => DATA_EXTENSIONS.some((ext) => n.endsWith(ext));
const isDiagramName = (n: string) => DIAGRAM_EXTENSIONS.some((ext) => n.endsWith(ext));

/**
 * The one classifier for root files. It returns the artifact kind of a root filename, or null if the
 * file is not an artifact. count, search, and discover all go through it, so a new root kind cannot slip
 * past one of them.
 */
function rootKind(nameLower: string): RootKind | null {
  if (isTasksName(nameLower)) return "tasks";
  if (isMarkdownName(nameLower)) return "markdown";
  if (isDataName(nameLower)) return "data";
  if (isDiagramName(nameLower)) return "diagram";
  return null;
}

/**
 * The filenames at the change root that pass `match`. It ignores dotfiles and directories, and sorts by
 * name. root-only and non-dotfile are the shared rules of every root-file listing, stated here once. It
 * does not descend: a file under a subdirectory is not an artifact.
 */
function listRootFiles(changePath: string, match: (nameLower: string) => boolean): string[] {
  if (!fs.existsSync(changePath)) return [];
  return fs
    .readdirSync(changePath, { withFileTypes: true })
    .filter((e) => e.isFile() && !e.name.startsWith(".") && match(e.name.toLowerCase()))
    .map((e) => e.name)
    .sort();
}

/**
 * The root artifact files with their kinds, in id-dedup precedence: markdown and tasks first, then the
 * extension-keeping kinds (data and diagram). This order lets spec.md keep the id "spec" and pushes
 * spec.json to spec-2. discoverArtifacts builds
 * from this list. The display order is a separate mtime sort in discoverArtifacts. It does one directory
 * read and partitions the entries by kind.
 */
export function rootArtifacts(changePath: string): { file: string; kind: RootKind }[] {
  if (!fs.existsSync(changePath)) return [];
  // The file-only, non-dotfile rules are the same ones listRootFiles states. They are kept here because
  // this walk also needs each entry's kind. Carry {file, kind} through the partition, so the kind is
  // computed once and not recomputed behind a non-null assertion after the sort.
  const md: { file: string; kind: RootKind }[] = [];
  const data: { file: string; kind: RootKind }[] = [];
  for (const e of fs.readdirSync(changePath, { withFileTypes: true })) {
    if (!e.isFile() || e.name.startsWith(".")) continue;
    const kind = rootKind(e.name.toLowerCase());
    if (!kind) continue;
    // data and diagram share the second bucket: both keep their extension in the title, and both yield
    // the bare stem to a markdown file of the same name (flow.md keeps `flow`, flow.mmd becomes flow-2).
    (kind === "data" || kind === "diagram" ? data : md).push({ file: e.name, kind });
  }
  // Default string order per bucket (matches the prior `.sort()` on names), so spec.md still precedes
  // spec.yaml within data and the id-dedup precedence in discoverArtifacts is unchanged.
  const byName = (a: { file: string }, b: { file: string }) =>
    a.file < b.file ? -1 : a.file > b.file ? 1 : 0;
  md.sort(byName);
  data.sort(byName);
  return [...md, ...data];
}

/**
 * The specs/ delta tree as a { topic, file } list, sorted by topic. It reads directory entries and tests
 * for each spec.md. It does not read content. hasSpecsTree, specsMtime, and the spec content read in
 * discoverArtifacts all derive from it, so the tree walk is written one time.
 */
export function listSpecFiles(changePath: string): { topic: string; file: string }[] {
  const specsDir = path.join(changePath, "specs");
  if (!fs.existsSync(specsDir) || !fs.statSync(specsDir).isDirectory()) return [];
  const out: { topic: string; file: string }[] = [];
  for (const topic of fs.readdirSync(specsDir).filter((n) => !n.startsWith("."))) {
    const file = path.join(specsDir, topic, "spec.md");
    if (fs.existsSync(file)) out.push({ topic, file });
  }
  return out.sort((a, b) => a.topic.localeCompare(b.topic));
}

/** True if specs/ holds at least one spec.md. It reads no content and returns on the first hit. This is
 *  the changes-list hot path (countArtifacts, called per change during a scan). So it does not build the
 *  full listSpecFiles list only to ask `.length > 0`. It walks the same specs/<topic>/spec.md shape as
 *  listSpecFiles. Keep the two in step. */
function hasSpecsTree(changePath: string): boolean {
  const specsDir = path.join(changePath, "specs");
  if (!fs.existsSync(specsDir) || !fs.statSync(specsDir).isDirectory()) return false;
  for (const topic of fs.readdirSync(specsDir)) {
    if (topic.startsWith(".")) continue;
    if (fs.existsSync(path.join(specsDir, topic, "spec.md"))) return true;
  }
  return false;
}

/** The sort time of the specs artifact: the newest mtime of every spec.md file (0 if none). */
export function specsMtime(changePath: string): number {
  let newest = 0;
  for (const { file } of listSpecFiles(changePath)) {
    const m = fs.statSync(file).mtimeMs;
    if (m > newest) newest = m;
  }
  return newest;
}

/**
 * The root files that count as a searchable artifact (markdown, tasks, data, and diagram), sorted by
 * name. A diagram file is indexed as the text it holds, not as the picture drawn from it: the labels an
 * author writes into a diagram are frequently the only place a term appears. The
 * specs delta tree is not here: it shows in the Specs tab, but its content does not go into search.
 *
 * It is name-sorted across all kinds, which is *not* the corpus order — search builds its documents from
 * `rootArtifacts` (markdown and tasks before data) via `collectSearchDocuments`, so that a change matched
 * by name alone is previewed from prose rather than from raw YAML. This function no longer has an in-repo
 * caller; it stays because it is a published export of @spekjs/core.
 */
export function listChangeArtifactFiles(changePath: string): string[] {
  return rootArtifacts(changePath)
    .map((a) => a.file)
    .sort();
}

/**
 * The root markdown filenames of a change (root *.md, including tasks.md). This is an existing published
 * export of @spekjs/core, kept for outside consumers. The app uses listChangeArtifactFiles instead.
 */
export function listChangeMarkdownFiles(changePath: string): string[] {
  return listRootFiles(changePath, isMarkdownName);
}

/**
 * The artifact count (the root artifact files, plus 1 for a non-empty specs tree). ChangeInfo uses it on
 * the changes-list hot path. It reads no content, and returns 0 for a missing changePath.
 */
export function countArtifacts(changePath: string): number {
  return rootArtifacts(changePath).length + (hasSpecsTree(changePath) ? 1 : 0);
}

/**
 * The newest mtime of the change directory: the maximum over the root artifact files and the specs tree.
 * The cross-worktree election uses it to find which copy was edited last.
 */
export function changeDirMtime(changePath: string): number {
  let newest = specsMtime(changePath);
  for (const { file } of rootArtifacts(changePath)) {
    const m = fs.statSync(path.join(changePath, file)).mtimeMs;
    if (m > newest) newest = m;
  }
  return newest;
}

File: packages/intellij/src/main/kotlin/com/spek/intellij/core/ArtifactFiles.kt
package com.spek.intellij.core

import java.io.File

/**
 * Aligned with @spekjs/core's artifact-files.ts. The filesystem view of one change directory: which files
 * are artifacts, their kinds, their names, and their mtimes. Everything here reads directory entries and
 * stats only. It does not read file content. ArtifactDiscovery.discover builds the ChangeArtifact objects
 * on top of this list.
 */
object ArtifactFiles {

    // internal, not private: WatchPolling builds its snapshot's watched-extension set (`.md` + these)
    // from this one source. A new data extension then cannot land here and be silently ignored there.
    internal val DATA_EXTENSIONS = listOf(".yaml", ".yml", ".json")

    // internal for the same reason as DATA_EXTENSIONS: WatchPolling builds its watched-extension set from
    // these lists rather than a third copy. A diagram is its own kind, not another data extension — `data`
    // means "show this file's text", `diagram` means "show what this file describes".
    internal val DIAGRAM_EXTENSIONS = listOf(".mmd", ".mermaid")

    /** The kind of a root-level artifact file. specs is not here: it is a tree, not a root file. */
    enum class RootKind { TASKS, MARKDOWN, DATA, DIAGRAM }

    private fun isTasksName(n: String) = n == "tasks.md"
    private fun isMarkdownName(n: String) = n.endsWith(".md")
    private fun isDataName(n: String) = DATA_EXTENSIONS.any { n.endsWith(it) }
    private fun isDiagramName(n: String) = DIAGRAM_EXTENSIONS.any { n.endsWith(it) }

    /**
     * The one classifier for root files. It returns the artifact kind of a root filename, or null if the
     * file is not an artifact. count, search, and discover all go through it, so a new root kind cannot
     * slip past one of them.
     */
    private fun rootKind(nameLower: String): RootKind? = when {
        isTasksName(nameLower) -> RootKind.TASKS
        isMarkdownName(nameLower) -> RootKind.MARKDOWN
        isDataName(nameLower) -> RootKind.DATA
        isDiagramName(nameLower) -> RootKind.DIAGRAM
        else -> null
    }

    /**
     * The root artifact files with their kinds, in id-dedup precedence: markdown and tasks first, then
     * the extension-keeping kinds (data and diagram). This order lets spec.md keep the id "spec" and
     * pushes spec.json to spec-2. discover builds
     * from this list. The display order is a separate mtime sort in discover. It does one directory read
     * and partitions the entries by kind.
     */
    fun rootArtifacts(changeDir: File): List<Pair<File, RootKind>> {
        // Carry (file, kind) through the partition so the kind is computed once, not recomputed behind a
        // !! after the sort.
        val md = mutableListOf<Pair<File, RootKind>>()
        val data = mutableListOf<Pair<File, RootKind>>()
        changeDir.listFiles()?.forEach { f ->
            if (!f.isFile || f.name.startsWith(".")) return@forEach
            when (val kind = rootKind(f.name.lowercase())) {
                // data and diagram share the second bucket: both keep their extension in the title, and
                // both yield the bare stem to a markdown file of the same name.
                RootKind.DATA, RootKind.DIAGRAM -> data.add(f to kind)
                RootKind.TASKS, RootKind.MARKDOWN -> md.add(f to kind)
                null -> {}
            }
        }
        md.sortBy { it.first.name }
        data.sortBy { it.first.name }
        return md + data
    }

    /**
     * The specs/ delta tree as a (topic, file) list, sorted by topic. It reads directory entries and tests
     * for each spec.md. It does not read content. hasSpecsTree, specsMtime, and the spec content read in
     * ArtifactDiscovery.discover all derive from it, so the tree walk is written one time.
     */
    fun listSpecFiles(changeDir: File): List<Pair<String, File>> {
        val specsDir = File(changeDir, "specs")
        if (!specsDir.isDirectory) return emptyList()
        return specsDir.listFiles()
            ?.filter { it.isDirectory && !it.name.startsWith(".") }
            ?.mapNotNull { topicDir ->
                val specFile = File(topicDir, "spec.md")
                if (specFile.exists()) topicDir.name to specFile else null
            }
            ?.sortedBy { it.first }
            ?: emptyList()
    }

    /** True if specs/ holds at least one spec.md. It reads no content and returns on the first hit. This
     *  is the changes-list hot path (count, called per change during a scan). So it does not build the full
     *  listSpecFiles list only to ask isNotEmpty(). It walks the same specs/<topic>/spec.md shape. */
    private fun hasSpecsTree(changeDir: File): Boolean {
        val specsDir = File(changeDir, "specs")
        if (!specsDir.isDirectory) return false
        return specsDir.listFiles()
            ?.any { it.isDirectory && !it.name.startsWith(".") && File(it, "spec.md").exists() }
            ?: false
    }

    /** The sort time of the specs artifact: the newest mtime of every spec.md file (0 if none). */
    fun specsMtime(changeDir: File): Long =
        listSpecFiles(changeDir).maxOfOrNull { (_, file) -> file.lastModified() } ?: 0L

    /**
     * The root files that count as a searchable artifact (markdown, tasks, data, and diagram), sorted by
     * name. A diagram is indexed as the text it holds, not the picture drawn from it. The
     * SearchService index shares this list, so any tab that comes from a root file is indexed. The specs
     * delta tree is not here: its content does not go into search.
     */
    fun artifactFiles(changeDir: File): List<File> =
        rootArtifacts(changeDir).map { it.first }.sortedBy { it.name }

    /** The artifact count (the root artifact files, plus 1 for a non-empty specs tree). It reads no
     *  content, and returns 0 for a missing changeDir. */
    fun count(changeDir: File): Int =
        rootArtifacts(changeDir).size + if (hasSpecsTree(changeDir)) 1 else 0
}
